File: bench/dispatch.fs
\ dispatch.fs — Phase 0.3 DISPATCH-bound microbench (the fair gate shape).
\ Per byte: load + add + 3x xorshift = ~10 cheap dispatched ops. gforth pays NEXT
\ per op; native folds each `dup lshift xor` into one shifted-EOR. Same hash both
\ ways (seed 1, buf[i]=i&0xff). Run: gforth dispatch.fs -e "BENCH bye".
65536 constant BUFLEN
15000 constant PASSES
create BUF BUFLEN allot
variable RES

: FILL   BUFLEN 0 ?do  i 255 and  BUF i + c!  loop ;

: HASH ( -- h )
  1
  PASSES 0 ?do
    BUFLEN 0 ?do
      BUF i + c@  +
      dup 13 lshift xor
      dup  7 rshift xor
      dup 17 lshift xor
    loop
  loop ;

: BENCH ( -- )
  FILL
  utime  HASH RES !  utime
  2swap d- d>s
  ." us=" .  ." result=" RES @ 255 and .  cr ;

File: bench/dispatch.c
/* dispatch.c — the dispatch-bound byte-mix as C, for an LLVM (clang -O3)
   baseline. Same algorithm/seed as dispatch.s. This is the bar a native Forth
   backend must rival. */
#include <stdint.h>
static uint8_t buf[65536];
int main(void){
    for (int i=0;i<65536;i++) buf[i]=i&0xff;
    uint64_t h=1;
    for (int p=0;p<15000;p++)
        for (int i=0;i<65536;i++){
            h += buf[i];
            h ^= h<<13; h ^= h>>7; h ^= h<<17;
        }
    return (int)(h & 0xff);
}
